libcudacxx/include/cuda/std/detail/libcxx/include/optional

```diff
@@ -1397,26 +1457,312 @@ operator>=(const _Tp& __v, const optional<_Up>& __x)
+    is_constructible, _Tp&, const _Up) _CCCL_AND(!__from_temporary<const _Up>))
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr optional(const optional<_Up>&& __u) noexcept(
+    noexcept(static_cast<_Tp&>(_CUDA_VSTD::declval<const _Up>())))
+      : __value_(__u.has_value() ? _CUDA_VSTD::addressof(static_cast<_Tp&>(__u.value())) : nullptr)
+  {}
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES(__from_temporary<const _Up>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional(const optional<_Up>&&) = delete;
+
+  _CCCL_HIDE_FROM_ABI constexpr optional& operator=(const optional&) noexcept = default;
+  _CCCL_HIDE_FROM_ABI constexpr optional& operator=(optional&&) noexcept      = default;
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional& operator=(nullopt_t) noexcept
+  {
+    __value_ = nullptr;
+    return *this;
+  }
+
+  _CCCL_TEMPLATE(class _Up = _Tp)
+  _CCCL_REQUIRES((!__is_std_optional<decay_t<_Up>>::value) _CCCL_AND _CCCL_TRAIT(is_constructible, _Tp&, _Up))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional&
+  operator=(_Up&& __u) noexcept(noexcept(static_cast<_Tp&>(_CUDA_VSTD::forward<_Up>(__u))))
+  {
+    __value_ = _CUDA_VSTD::addressof(__u);
+    return *this;
+  }
+
+  _CCCL_TEMPLATE(class _Up = _Tp)
+  _CCCL_REQUIRES((!__is_std_optional<decay_t<_Up>>::value) _CCCL_AND _CCCL_TRAIT(is_constructible, _Tp&, _Up)
+                   _CCCL_AND(!__from_temporary<_Up>))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp&
+  emplace(_Up&& __u) noexcept(noexcept(static_cast<_Tp&>(_CUDA_VSTD::forward<_Up>(__u))))
+  {
+    __value_ = _CUDA_VSTD::addressof(static_cast<_Tp&>(_CUDA_VSTD::forward<_Up>(__u)));
+    return *__value_;
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr void swap(optional& __rhs) noexcept
+  {
+    return _CUDA_VSTD::swap(__value_, __rhs.__value_);
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp* operator->() const noexcept
+  {
+    _CCCL_ASSERT(__value_ != nullptr, "optional operator-> called on a disengaged value");
+    return __value_;
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp& operator*() const noexcept
+  {
+    _CCCL_ASSERT(__value_ != nullptr, "optional operator* called on a disengaged value");
+    return *__value_;
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr explicit operator bool() const noexcept
+  {
+    return __value_ != nullptr;
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr bool has_value() const noexcept
+  {
+    return __value_ != nullptr;
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp& value() const noexcept
+  {
+    if (__value_ != nullptr)
+    {
+      return *__value_;
+    }
+    else
+    {
+      __throw_bad_optional_access();
+    }
+  }
+
+  template <class _Up>
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr remove_cvref_t<_Tp> value_or(_Up&& __v) const
+  {
+    static_assert(_CCCL_TRAIT(is_copy_constructible, _Tp), "optional<T>::value_or: T must be copy constructible");
+    static_assert(_CCCL_TRAIT(is_convertible, _Up, _Tp), "optional<T>::value_or: U must be convertible to T");
```

Comment:
Nit: should the assert messages say `optional<T&>` instead of `optional<T>`?

Reply:
they definitely should 👍